near-sdk/src/store/tree_map/mod.rs

```diff
@@ -0,0 +1,2059 @@
+mod entry;
+mod impls;
+mod iter;
+
+pub use entry::Entry;
+pub use iter::{Iter, IterMut, Keys, Range, RangeMut, Values, ValuesMut};
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::borrow::Borrow;
+use std::ops::RangeBounds;
+
+use super::lookup_map as lm;
+use crate::crypto_hash::{CryptoHasher, Sha256};
+use crate::store::free_list::{FreeList, FreeListIndex};
+use crate::store::LookupMap;
+use crate::{env, IntoStorageKey};
+
+type NodeAndIndex<'a, K> = (FreeListIndex, &'a Node<K>);
+
+fn expect<T>(val: Option<T>) -> T {
+    val.unwrap_or_else(|| env::abort())
+}
+
+/// TreeMap based on AVL-tree
+///
+/// Runtime complexity (worst case):
+/// - `get`/`contains_key`:     O(1) - LookupMap lookup
+/// - `insert`/`remove`:        O(log(N))
+/// - `min`/`max`:              O(log(N))
+/// - `above`/`below`:          O(log(N))
+/// - `range` of K elements:    O(Klog(N))
```

Comment:
Similar thing here - if we make iterators actually hold `Node` instances, we would be able to answer `range` queries with a better complexity of `O(K + log(N))`. Every time you need `next` element, you would just follow a deterministic algorithm of finding the next/previous value in a binary tree. This might be a little more worthwhile than my last point, but I leave it to your judgement.

I am not 100% sure on my assessment of complexity as each next-element-lookup takes `O(log(N))`, so if we judge the complexity naively it would still be `O(Klog(N))`. However, if we actually do amortized analysis here, I think in a BST with inner nodes it should be `O(K + log(N))`. I can uncover some existing formal proof on this if you want or, if not, probably can come up with one myself.

Reply:
It's definitely possible and should be done, but requires a decent refactor as with the current impl we don't have a way to traverse up to the parent for these operations. Perhaps like the previous point we should implement this optimization in a future PR?

Reply:
Sure